### ISSUE TITLE
iOS notifications on coldstart not working.

### ISSUE THREAD

**Issue by emc2spider** [AppGyver/PushNotifications#9]
When an iOS notification arrives on background, it is received as `coldstart=true` and so it's processed like this, which to the intent of my application means no problem. But when real coldstart notifications arrive, meaning the app is not running at all so the notification has to launch the app from scratch, it seems that the notification handler callback is never executed or at least I don't get to see the expected effects from the notification. Debugging with Safari is useless because the console doesn't show previous messages, only those that arrive after I opened it.

**Comment by harsaharsa** [comment]
Yup, this is a bug/missing feature. The way to fix this is to implement the `applicationWillFinishLaunchingWithOptions` method that will put the received notification into the message queue from the launch options dictionary.

Or, more robustly, refactor the code to use the `didReceiveRemoteNotification:fetchCompletionHandler` method instead of the current `didReceiveRemoteNotification` to handle the background push notification and save it in a native queue. These are on the backlog (though PRs are welcome too).

Apple reference:
https://developer.apple.com/library/ios/documentation/UIKit/Reference/UIApplicationDelegate_Protocol/index.html#//apple_ref/occ/intfm/UIApplicationDelegate/application:didReceiveRemoteNotification:

**Comment by emc2spider** [comment]
Hey, any updates on this one?

**Comment by harsaharsa** [comment]
I'm looking into this this week, stay tuned for updates!

**Comment by harsaharsa** [comment]
It seems that true coldstart notifications (i.e. app is not runninG) ARE handled correctly and sent to the message queue, but only if the notification itself is used to open the app.

For notifications to get processed even if the app is not opened via the push notification alert, you need to add `remote-notifications` to the list of `UIBackgroundModes` in Info.plist. You can do this via `plugin.xml`.

**Comment by emc2spider** [comment]
I'm giving this a try, thanks @harsaharsa!

**Comment by rufiodev** [comment]
Hi, I would like to try this solution but I'm not quite sure how to go about it.

I believe I have to fork this plugin and add some info to plugin.xml, right? If this is done, I can get a callback to the push notification of the app when clicking on the notification in the bar.

Did anybody get the cold start callback to work? Thanks in advance for any explanation you can give!

**Comment by harsaharsa** [comment]
I don't remember the syntax 100% off the top of my head, but it should be something like
```
<config-file target="*-Info.plist" parent="UIBackgroundModes">
  <string>remote-notifications</string>
</config-file>
```

In your `plugin.xml`. You can view the end result by renaming the .ipa to .zip, extracting it and using Preview to view the Info.plist file in the archive.

**Comment by rufiodev** [comment]
Thanks for your quick reply and example @harsaharsa 

The code you send crashes sadly my app immediately after opening it. It seems it gets included correctly in the info.plist:
<img width="665" alt="screen shot 2015-07-17 at 14 55 54" src="https://cloud.githubusercontent.com/assets/13215722/8747551/790bd0bc-2c94-11e5-80a5-16fe3e2981b2.png">

I found that when I add an `<array>` around the string my app does not crash.
```
<config-file target="*-Info.plist" parent="UIBackgroundModes">
   <array>
        <string>remote-notifications</string>
   </array>
</config-file>
```

In this case it also gets included in the info.plist, seems ok to me:
<img width="774" alt="screen shot 2015-07-17 at 15 02 21" src="https://cloud.githubusercontent.com/assets/13215722/8747638/ef18c864-2c94-11e5-8261-2543728a8976.png">

But after firing a push notification when the app is closed, I still get no callback with data in the app (I'm clicking on the push notification).

What am I doing wrong?

**Comment by rufiodev** [comment]
Ok, to continue directly on the original question.

This function does not seem to get called on  __iOS cold start__:
```
pushNotification.onMessageInBackground(
	messageInBackgroundHandler,
	errorHandler
);
```

Not when I click on the app itself and also not when I click on the notification in the notification bar. Because of this I am unable to send users to a specific page in the app.

I tried to set a setTimeout around the function with 400ms, as I read somewhere else, but that sadly doesn't do the trick.

Any help anybody can give me is greatly appreciated. @emc2spider, did you found a solution to this?

**Comment by 123AndGo** [comment]
I'm having the same problem. I can't receive the notification's content when the app is closed neither when it opens by clicking the notification nor when clicking the icon :(

**Comment by harsaharsa** [comment]
Sorry for being silent on this – we are going to have a push notification sprint sometime soonish to fix this and other issues for good, but there's other stuff in the immediate backlog ATM.

**Comment by emc2spider** [comment]
Nobody has mentioned this, but I see on the comments you're typing `remote-notifications` in plural, it should be singular `remote-notification` maybe it solves the whole problem? My project's been working fine for a while now, though I haven't upgraded this particular plugin lately :)

**Comment by gaj-rscpp** [comment]
Hey guys, I am having the same issue. I tried the fix suggested above but still no luck.  Is there anyway updates on this issue?

**Comment by edskeizer** [comment]
Any progress on this guys? Still facing the same issue here...
